symbolic-base/src/ZkFold/Symbolic/Data/Input.hs

```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module ZkFold.Symbolic.Data.Input (
+    SymbolicInput (..)
+) where
+
+import           Control.Monad.Representable.Reader (Rep)
+import           Data.Foldable                      (foldlM)
+import           Data.Functor
+import           Data.Functor.Rep                   (Representable)
+import           Data.Ord                           (Ord)
+import           Data.Traversable                   (for)
+import           Data.Type.Equality                 (type (~))
+import           Data.Typeable                      (Proxy (..))
+import           GHC.Generics                       (Par1 (..))
+import           Prelude                            (const, ($))
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.Number   (KnownNat)
+import           ZkFold.Base.Data.ByteString        (Binary)
+import           ZkFold.Base.Data.Vector
+import           ZkFold.Symbolic.Class
+import           ZkFold.Symbolic.Data.Bool
+import           ZkFold.Symbolic.Data.ByteString
+import           ZkFold.Symbolic.Data.Class
+import           ZkFold.Symbolic.Data.Combinators
+import           ZkFold.Symbolic.Data.UInt
+import           ZkFold.Symbolic.MonadCircuit
+
+
+-- | A class for Symbolic input.
+class
+    ( SymbolicData d
+    , Support d ~ Proxy (Context d)
+    , Representable (Layout d)
+    , Binary (Rep (Layout d))
+    , Ord (Rep (Layout d))
+    ) => SymbolicInput d where
+    isValid :: d -> Bool (Context d)
+
+
+instance Symbolic c => SymbolicInput (Bool c) where
+  isValid (Bool b) = Bool $ fromCircuitF b $
+      \(Par1 v) -> Par1 <$> newAssigned (($ v) * (one - ($ v))) -- проверка, что i -- это 0 или 1
```

Comment:
This is not a check that $b \in [0; 1]$, sadly: this is $0$ when $b \in [0; 1]$ and something else otherwise.
A correct check would look like
```haskell
\(Par1 v) -> do
  u <- newAssigned (\x -> x v * (one - x v))
  fst <$> runInvert (Par1 u)
```